### ISSUE TITLE
Quantile Regression does not split data point counts correctly above/below the regression line

### ISSUE THREAD

**Issue by josepm** [statsmodels/statsmodels#6928]
#### Describe the bug

When running a QR with say q = 0.75, one expects 25% of the data points to be above the regression line. Sometimes the QR algorithm fails to do that.

#### Code Sample, a copy-pastable example if possible
 Attaching a notebook with examples that shows the problem.

[QR_ipynb.zip](https://github.com/statsmodels/statsmodels/files/4990277/QR_ipynb.zip)

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

<details>

[paste the output of ``import statsmodels.api as sm; sm.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
Python: 3.7.4.final.0
OS: Darwin 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

statsmodels
===========

Installed: 0.11.1 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/statsmodels)

Required Dependencies
=====================

cython: 0.29.13 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/Cython)
numpy: 1.17.2 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/numpy)
scipy: 1.4.1 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/scipy)
pandas: 1.0.4 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/pandas)
    dateutil: 2.8.0 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/dateutil)
patsy: 0.5.1 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/patsy)

Optional Dependencies
=====================

matplotlib: 3.2.1 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/matplotlib)
    backend: MacOSX 
cvxopt: Not installed
joblib: 0.15.1 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/joblib)

Developer Tools
================

IPython: 7.8.0 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/IPython)
    jinja2: 2.11.2 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/jinja2)
sphinx: 2.2.0 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/sphinx)
    pygments: 2.4.2 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages/pygments)
pytest: 5.2.1 (/Users/josep_ferrandiz/.anaconda/Anaconda3-2019.10/lib/python3.7/site-packages)
virtualenv: Not installed

</details>

**Comment by bashtage** [comment]
Have you checked both the first order condition? When you have a nonconstant X regressor the split is not necessarily exact.

**Comment by bashtage** [comment]
I checked the FOC when q=0.5, since this is the easy case, and statsmodels generally does a better job than scikit-learn:

statsmodels
```
xcols: ['income2'] expected count: 118.0 actual count: 101 result: ERROR foc: [-0.14656052]
xcols: ['const'] expected count: 118.0 actual count: 117 result: OK foc: [-0.00425532]
xcols: ['income'] expected count: 118.0 actual count: 134 result: ERROR foc: [4.58464664]
xcols: ['const', 'income'] expected count: 118.0 actual count: 118 result: OK foc: [4.25531915e-03 4.40608132e+00]
xcols: ['income', 'income2'] expected count: 118.0 actual count: 116 result: OK foc: [-1.28578816 -0.05750652]
xcols: ['const', 'income', 'income2'] expected count: 118.0 actual count: 118 result: OK foc: [ 0.00425532 -0.26981145  0.07780122]
```

scikit-learn
```
xcols: ['income2'] expected count: 118.0 actual count: 118 result: OK foc: [0.50858669]
xcols: ['const'] expected count: 118.0 actual count: 117 result: OK foc: [0.]
xcols: ['income'] expected count: 118.0 actual count: 118 result: OK foc: [34.70815243]
xcols: ['const', 'income'] expected count: 118.0 actual count: 118 result: OK foc: [4.25531915e-03 3.47081524e+01]
xcols: ['income', 'income2'] expected count: 118.0 actual count: 118 result: OK foc: [34.70815243  0.50858669]
xcols: ['const', 'income', 'income2'] expected count: 118.0 actual count: 118 result: OK foc: [4.25531915e-03 3.47081524e+01 5.08586695e-01]
```

All of the foc values should be approx 0.

**Comment by bashtage** [comment]
Here is the updated code:

```
#!/usr/bin/env python
# coding: utf-8

# with QR(q) we should have (1 - q)% of the data points above the quantile regression line
# unfortunately this is not the case for the statsmodel implementation
# based on https://www.statsmodels.org/dev/examples/notebooks/generated/quantile_regression.html
import pandas as pd
import statsmodels.api as sm
import numpy as np

data = sm.datasets.engel.load_pandas().data
data = sm.add_constant(data)            # add 'const' column with all 1's
data['income2'] = np.sqrt(data['income'].values)

def get_smcnt(q, data, xcols):
    # count points above the regression line dor the statsmodel implementation
    # prints OK if counts match theoretical value, otherwise prints ERROR
    # q: quantile
    # data: DF with X and y
    # xcols: cols used for X
    y = data['foodexp'].values
    X = data[xcols].values
    qmdl = sm.QuantReg(y, X)
    res = qmdl.fit(q)
    p = np.asarray(res.params)
    data['foodexp_pred'] = qmdl.predict(res.params, exog=X)                          # in sample prediction
    foc = None
    if q == 0.5:
        err = y - data['foodexp_pred']
        foc = np.mean(np.sign(err)[:,None] * X, axis=0)
    w = data.apply(lambda x: 1 if x['foodexp'] > x['foodexp_pred'] else 0, axis=1)  # 1 if above the QR line
    p = (1 - q) * len(data)   # 1 - q data fraction
    expected = np.round(p, 0)
    actual = w.sum()
    result = 'ERROR' if np.abs((expected - actual) / actual) > 0.05 else 'OK'
    print('xcols: ' + str(xcols) + 
          ' expected count: ' + str(expected) + 
          ' actual count: ' + str(actual) + ' result: ' + result + " foc: " + str(foc))
    


# with the statsmodels QR, some counts are off
for xcols in [['income2'], ['const'], ['income'], 
              ['const', 'income'], ['income', 'income2'], ['const', 'income', 'income2' ]]:
    _ = [get_smcnt(q, data, xcols) for q in [0.25, 0.5, 0.75]]


# try scikit-learn QR implementation
from sklearn.ensemble import GradientBoostingRegressor
def get_skcnt(q, data, xcols):
    # count points above the regression line dor the scikit-learn QR implementation
    # prints OK if counts match theoretical value, otherwise prints ERROR
    # q: quantile
    # data: DF with X and y
    # xcols: cols used for X
    y = data['foodexp'].values
    X = data[xcols].values
    reg = GradientBoostingRegressor(loss='quantile', alpha=q, random_state=0)
    reg.fit(X, y)
    data['foodexp_pred'] = reg.predict(X)  # in sample prediction
    foc = None
    if q == 0.5:
        err = y - data['foodexp_pred']
        foc = np.mean(np.sign(err)[:,None] * X, axis=0)  
    w = data.apply(lambda x: 1 if x['foodexp'] > x['foodexp_pred'] else 0, axis=1)  # 1 if above the QR line
    p = (1 - q) * len(data)                                                         # 1 - q data fraction
    expected = np.round(p, 0)
    actual = w.sum()
    result = 'ERROR' if np.abs((expected - actual) / actual) > 0.05 else 'OK'
    print('xcols: ' + str(xcols) + 
          ' expected count: ' + str(expected) + 
          ' actual count: ' + str(actual) + ' result: ' + result + " foc: " + str(foc))


# with the scikit-learn QR all counts are correct
for xcols in [['income2'], ['const'], ['income'], 
              ['const', 'income'], ['income', 'income2'], ['const', 'income', 'income2' ]]:
    _ = [get_skcnt(q, data, xcols) for q in [0.25, 0.5, 0.75]]


```

**Comment by bashtage** [comment]
Please don't post notebooks since they are a pain to quickly access. See #6913.

**Comment by bashtage** [comment]
You should probably report this to scikit learn.

**Comment by josef-pkt** [comment]
In general, you have to take into account zero residuals.
Because the objective function is not differentiable at zero, there is a fudge term around zero in the irls algorithm that is used for the estimation. So the count should include only points away from a neighborhood of zero.

@bashtage In your previous comment with the results, sklearn is doing better than statsmodels, AFAICS

**Comment by bashtage** [comment]
You missed the e+01 on the mean score. It is doing a very poor job fitting qr when income is present.

**Comment by bashtage** [comment]
sm
0.00425532 -0.26981145  0.07780122
SKL
. 00425531915 34.7081524 0.508586695

SKL appears to be broken given that is the average score here.

**Comment by bashtage** [comment]
I did log food exp:

https://gist.github.com/bashtage/e7c5f7e66e3e2c38fbfdd3dfbdcaf658

**Comment by bashtage** [comment]
SKL definitely seems to perform worse in terms of the quantile regression scores:

![image](https://user-images.githubusercontent.com/5585221/88958929-14433700-d299-11ea-974d-216e3a42767f.png)

**Comment by josef-pkt** [comment]
I just ran into something strange looking where I always had about half the residuals below zero independently which `q` I used,
until I realized that I did not have a constant in my exog.

I guess this is similar to OLS without a constant (regression through the origin). In that case the mean of the residuals cannot be zero,

In my simulated example I had a lot of symmetry, 10 exogs and large slope effects. So the slope effect was always dominating and without constant it could not shift the regression plane to the appropriate quantile.

The results in notebook linked by  @bashtage are similar where counts do not match up when no constant is in the regression.
(I did not see that before)
